core/node/da_clients/src/eigen/verifier.rs

```diff
@@ -0,0 +1,553 @@
+            return Ok(latest);
+        }
+        Ok(latest - (self.cfg.settlement_layer_confirmation_depth as u64 - 1))
+    }
+
+    async fn call_batch_id_to_metadata_hash(
+        &self,
+        blob_info: BlobInfo,
+    ) -> Result<Vec<u8>, VerificationError> {
+        let context_block = self.get_context_block().await?;
+
+        let func_selector =
+            ethabi::short_signature("batchIdToBatchMetadataHash", &[ParamType::Uint(32)]);
+        let mut data = func_selector.to_vec();
+        let mut batch_id_vec = [0u8; 32];
+        U256::from(blob_info.blob_verification_proof.batch_id).to_big_endian(&mut batch_id_vec);
+        data.append(batch_id_vec.to_vec().as_mut());
+
+        let call_request = CallRequest {
+            to: Some(
+                H160::from_str(&self.cfg.svc_manager_addr)
+                    .map_err(|_| VerificationError::ServiceManagerError)?,
+            ),
+            data: Some(zksync_basic_types::web3::Bytes(data)),
+            ..Default::default()
+        };
+
+        let res = self
+            .signing_client
+            .as_ref()
+            .call_contract_function(
+                call_request,
+                Some(BlockId::Number(BlockNumber::Number(context_block.into()))),
+            )
+            .await
+            .map_err(|_| VerificationError::ServiceManagerError)?;
+
+        Ok(res.0.to_vec())
+    }
+
+    /// Verifies the certificate batch hash
+    pub async fn verify_batch(&self, blob_info: BlobInfo) -> Result<(), VerificationError> {
+        let expected_hash = self
+            .call_batch_id_to_metadata_hash(blob_info.clone())
+            .await?;
+
+        if expected_hash == vec![0u8; 32] {
+            return Err(VerificationError::EmptyHash);
+        }
+
+        let actual_hash = self.hash_batch_metadata(
+            blob_info
+                .blob_verification_proof
+                .batch_medatada
+                .batch_header,
+            blob_info
+                .blob_verification_proof
+                .batch_medatada
+                .signatory_record_hash,
+            blob_info
+                .blob_verification_proof
+                .batch_medatada
+                .confirmation_block_number,
+        );
+
+        if expected_hash != actual_hash {
+            return Err(VerificationError::DifferentHashes);
+        }
+        Ok(())
+    }
+
+    fn decode_bytes(&self, encoded: Vec<u8>) -> Result<Vec<u8>, String> {
+        // Ensure the input has at least 64 bytes (offset + length)
+        if encoded.len() < 64 {
+            return Err("Encoded data is too short".to_string());
+        }
+
+        // Read the offset (first 32 bytes)
+        let offset = {
+            let mut offset_bytes = [0u8; 32];
+            offset_bytes.copy_from_slice(&encoded[0..32]);
+            usize::from_be_bytes(
+                offset_bytes[24..32]
+                    .try_into()
+                    .map_err(|_| "Offset is too large")?,
```

Comment:
The error message doesn't make sense here; the conversion is infallible on 64-bit architectures because the input slice has correct length. I'd still convert to `u64` first and then to `usize`.

Reply:
Fixed here https://github.com/lambdaclass/zksync-era/pull/375